experimental/dds/tree2/src/core/tree/visitDelta.ts

```diff
@@ -136,41 +124,66 @@ function fixedPointVisitOfRoots(
  * TODO#5481: update the DetachedFieldIndex instead of moving the nodes around.
  *
  * @param rootTransfers - The transfers to perform.
- * Entries are removed as they are performed.
  * @param mapToUpdate - A map to update based on the transfers being performed.
  * @param detachedFieldIndex - The index to update based on the transfers being performed.
  * @param visitor - The visitor to inform of the transfers being performed.
  */
 function transferRoots(
-	rootTransfers: RootTransfers,
+	rootTransfers: readonly Delta.DetachedNodeRename[],
 	mapToUpdate: Map<ForestRootId, unknown>,
 	detachedFieldIndex: DetachedFieldIndex,
 	visitor: DeltaVisitor,
 ): void {
-	while (rootTransfers.size > 0) {
-		const priorSize = rootTransfers.size;
-		for (const [oldRootId, { oldId, newId }] of rootTransfers) {
-			if (detachedFieldIndex.tryGetEntry(newId) !== undefined) {
+	type AtomizedNodeRename = Omit<Delta.DetachedNodeRename, "count">;
+	let nextBatch = rootTransfers.flatMap(({ oldId, newId, count }) => {
+		const atomized: AtomizedNodeRename[] = [];
+		// It's possible for a detached node to be revived transiently such that it ends up back in the same detached field.
+		// Making such a transfer wouldn't just be inefficient, it would lead us to mistakenly think we have moved all content
+		// out of the source detached field, and would lead us to delete the tree index entry for that source detached field.
+		// This would effectively result in the tree index missing an entry for the detached field.
+		// This if statement prevents that from happening.
+		if (!areDetachedNodeIdsEqual(oldId, newId)) {
+			for (let i = 0; i < count; i += 1) {
+				atomized.push({ oldId: offsetDetachId(oldId, i), newId: offsetDetachId(newId, i) });
+			}
+		}
+		return atomized;
+	});
+	while (nextBatch.length > 0) {
+		const delayed: AtomizedNodeRename[] = [];
+		const priorSize = nextBatch.length;
+		while (nextBatch.length > 0) {
```

Comment:
The nested while-loops with the same condition is kind of confusing. Can we make the inner loop a for-loop instead?